source/material_model/rheology/friction_options.cc

```diff
@@ -0,0 +1,495 @@
+                                          + rate_and_state_parameter_a
+                                          * log((current_edot_ii * cellsize ) / steady_state_strain_rate[j]) + rate_and_state_parameter_b
+                                          * log((theta * steady_state_strain_rate[j] ) / critical_slip_distance));
+                  break;
+                }
+              else
+                {
+                  break;
+                }
+            }
+            default:
+            {
+              AssertThrow(false, ExcNotImplemented());
+              break;
+            }
+          }
+        return current_friction;
+      }
+
+      template <int dim>
+      ComponentMask
+      FrictionOptions<dim>::
+      get_theta_composition_mask(ComponentMask composition_mask) const
+      {
+        // Store which components to exclude during the volume fraction computation.
+
+        if (get_theta_in_use())
+          {
+            // this is the compositional field used for theta in rate-and-state friction
+            const int theta_position_tmp = this->introspection().compositional_index_for_name("theta");
+            composition_mask.set(theta_position_tmp,false);
+          }
+
+        return composition_mask;
+      }
+
+
+      template <int dim>
+      double
+      FrictionOptions<dim>::
+      compute_theta(const double theta_old,
+                    const double current_edot_ii,
+                    const double cellsize) const
+      {
+        // equation (7) from Sobolev and Muldashev 2017. Though here I had to add  "- theta_old"
+        // because I need the change in theta for reaction_terms
+        const double current_theta = critical_slip_distance / ( cellsize * current_edot_ii ) +
+                                     (theta_old - critical_slip_distance / ( cellsize * current_edot_ii))
+                                     * exp( - ((current_edot_ii * cellsize) * this->get_timestep()) / critical_slip_distance);
+        return current_theta;
+      }
+
+
+      template <int dim>
+      void
+      FrictionOptions<dim>::
+      compute_theta_reaction_terms(const int q,
+                                   const MaterialModel::MaterialModelInputs<dim> &in,
+                                   const std::vector<double> &volume_fractions,
+                                   const double min_strain_rate,
+                                   const double ref_strain_rate,
+                                   const bool use_elasticity,
+                                   const bool use_reference_strainrate,
+                                   const std::vector<double> &elastic_shear_moduli,
+                                   const double dte,
+                                   MaterialModel::MaterialModelOutputs<dim> &out) const
+      {
+        //cellsize is needed for theta and the friction angle
+        double cellsize = 1;
+        if (in.current_cell.state() == IteratorState::valid)
+          {
```

Comment:
at the moment this is fine for what I do, because the cells are cubic and motion is in y-direction. But I remember that we discussed that it should probably rather be cell size in the direction of most movement. Unfortunately I do not remember how I can specify this. Do you have an idea?

Reply:
To get the length of the element in the slip direction, you'd have to
1. determine whether you're in the fault
2. if so, get the velocity vector
3. get vectors along two non-parallel edges of the element (I'm foreseeing mesh deformation here)
4. from this compute the length of the element along the velocity vector.
For now this is too much. If you can add a comment here that explains your choice, then it's fine, e.g.
`For now, the used cells are non-deforming squares, so the edge length in the x-direction is representative of the cell size. TODO as the cell size is used to compute the slip velocity as cell_size * strain_rate, come up with a better representation of the slip length.`

Reply:
thanks!